homeassistant/components/kaleidescape/entity.py

```diff
@@ -39,8 +38,7 @@ def __init__(self, device: KaleidescapeDevice) -> None:
     async def async_added_to_hass(self) -> None:
         """Register update listener."""
 
-        @callback
-        def _update(event: str) -> None:
+        async def _update(event: str) -> None:
```

Comment:
This fixes some race conditions with unit testing. Sometimes events sent into the Kaleidescape library dispatcher were not making into the hass entity state before an assertion was made. This ensures `await hass.async_block_till_done()` works as intended in tests.

Reply:
Making this a coroutine is far less efficient as callback functions don't have to be scheduled in the event loop.  It would be better to make a test helper to patch the pypi library to wait for the callback to fire using https://docs.python.org/3/library/asyncio-sync.html

Reply:
Reverted the change, and fixed the issue in the pypi library here https://github.com/SteveEasley/pykaleidescape/commit/79130c0efbb82098a1a2d97ec271affff8811450. But would like feedback.

This was after more digging into the issue I realized the root of the issue was my library (inspired by the Heos integration pypi library) was using `self._loop.run_in_executor()` to send updates to hass. Swapping to `self._loop.call_soon()` seems to be how [hass core does](https://github.com/home-assistant/core/blob/dev/homeassistant/core.py#L446) @callback calls, so emulated that.

Do you think thats an acceptable fix? Or in that same method should I just be calling the target directly `target(*args)`?

Reply:
> It would be better to make a test helper to patch the pypi library to wait for the callback to fire

Yeah I tried something like that early on, and didn't help. The issue was the callback does get fired, and I can even block till it does (with helpers), but hass still was intermittently not updating the media_player entity until later. Which as pointed out above, was because the library callback was being fired in another thread via `loop.run_in_executor()`.

Reply:
I'm in flight so I can't dig through the library at the moment but if everything is running in the event loop there is no need to delay the callback with call_soon. It's preferable to run it right away as long as it's not cpu intensive instead of using call_soon the schedule it.

Reply:
Yeah I guess `call_soon()` is not providing any benefit since the target it's calling is not an awaitable anyway. In this case listeners (platform entity) is just calling `async_write_ha_state()`, so should be fast.

Reply:
This was resolved in https://github.com/SteveEasley/pykaleidescape/releases/tag/v1.0.1